pip/exceptions.py

```diff
@@ -39,13 +43,209 @@ class PreviousBuildDirError(PipError):
     """Raised when there's a previous conflicting build directory"""
 
 
-class HashMismatch(InstallationError):
-    """Distribution file hash values don't match."""
-
-
 class InvalidWheelFilename(InstallationError):
     """Invalid wheel filename."""
 
 
 class UnsupportedWheel(InstallationError):
     """Unsupported wheel."""
+
+
+# The recommended hash algo of the moment. Feel free to change this any time.
+FAVORITE_HASH = 'sha256'
+
+
+class HashErrors(InstallationError):
+    """Multiple HashError instances rolled into one for reporting"""
+
+    def __init__(self):
+        self.errors = []
+
+    def append(self, error):
+        self.errors.append(error)
+
+    def __str__(self):
+        lines = []
+        self.errors.sort(key=lambda e: e.order)
+        for cls, errors_of_cls in groupby(self.errors, lambda e: e.__class__):
+            lines.append(cls.head())
+            lines.extend(e.body() for e in errors_of_cls)
+        if lines:
+            return '\n'.join(lines)
+
+    def __nonzero__(self):
+        return bool(self.errors)
+
+    def __bool__(self):
+        return self.__nonzero__()
+
+
+class HashError(InstallationError):
+    """A failure to verify a package against known-good hashes
```

Comment:
Please break after the first triple quote to be consistent with the majority of the other docstrings.